server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -335,11 +339,13 @@ public TranslogReader closeIntoReader() throws IOException {
                         throw ex;
                     }
                     // If we reached this point, all of the buffered ops should have been flushed successfully.
-                    assert bufferedOps.size() == 0;
+                    assert buffer == null;
                     assert checkChannelPositionWhileHandlingException(totalOffset);
                     assert totalOffset == lastSyncedCheckpoint.offset;
                     if (closed.compareAndSet(false, true)) {
                         try {
+                            Releasables.closeWhileHandlingException(buffer);
+                            buffer = null;
```

Comment:
I struggle to see how this can do anything, given that we assert that `buffer` is `null` a few lines above and we retain `synchronized(this)` throughout?